### ISSUE TITLE
False positive invalid-overridden-method with decorator

### ISSUE THREAD

**Issue by darkclouder** [pylint-dev/pylint#8939]
### Bug description

I think I get false positives when I use a 

error.py:
```python
from abc import ABCMeta, abstractmethod
import asyncio
from contextlib import AbstractAsyncContextManager, asynccontextmanager

class Base(metaclass=ABCMeta):
    @abstractmethod
    def get(self) -> AbstractAsyncContextManager[str]:
        pass

class Implementation(Base):
    @asynccontextmanager
    async def get(self):
        yield "hello"

async def main():
    async with Implementation().get() as res:
        print(res)

if __name__ == "__main__":
    asyncio.run(main())
```

### Command used

```shell
pylint error.py
```

### Pylint output

```shell
************* Module error
error.py:12:4: W0236: Method 'get' was expected to be 'non-async', found it instead as 'async' (invalid-overridden-method)

------------------------------------------------------------------
Your code has been rated at 9.29/10 (previous run: 9.29/10, +0.00)
```

### Expected behavior

For it to pass because `asynccontextmanager` returns a sync function, which is then put in the place of `Implementation.get` and hence the overridden method matches the signature of `Base`.

### Pylint version

```shell
pylint 2.17.5
astroid 2.15.6
Python 3.11.4 | packaged by conda-forge | (main, Jun 10 2023, 18:10:28) [Clang 15.0.7 ]
```

### OS / Environment

macOS 13.5
Miniconda

**Comment by mickare** [comment]
Hey! I can confirm this bug.
Although ... it may be a duplicate of #5761 ?

**Comment by mickare** [comment]
If you change it for pylint then mypy will fail.

Here is a nice workaround to get both pylint and mypy working:
```python3
from abc import ABCMeta, abstractmethod
import asyncio
from contextlib import asynccontextmanager
from typing import AsyncIterator


class Base(metaclass=ABCMeta):
    @abstractmethod
    @asynccontextmanager
    async def get(self) -> AsyncIterator[str]:
        yield NotImplementedError()  # type: ignore


class Implementation(Base):
    @asynccontextmanager
    async def get(self) -> AsyncIterator[str]:
        yield "hello"


async def main():
    async with Implementation().get() as res:
        print(res)


if __name__ == "__main__":
    asyncio.run(main())
```

Reason: Python requires a `yield` keyword in a function to actually transform the function's signature to a generator or iterator. In abstract interface definitions this is missing and therefore the signature will be different.

This is a inconsistency and implicit behavior in the Python language itself and is a problem for all static type checkers and linters.
